### ISSUE TITLE
Reduce duplicate css/js in webpack output

### ISSUE THREAD

**Issue by ChristopherChudzicki** [mitodl/ocw-hugo-themes#1033]
We bundle the js/css assets that a user downloads with webpack. Currenlty, users download a few different "bundles". Mostly:
- visiting a course --> download `main` and `course_v2`
- visiting site homepage (ocw.mit.edu) --> download `main` and `www`

This currently results in users downloading a fair amount of duplicated code. A few examples are jquery, popper, and sentry:

<img width="1728" alt="Screenshot 2023-01-23 at 2 42 10 PM" src="https://user-images.githubusercontent.com/9010790/214135770-0d557ab4-1a75-4cee-aaba-3808f6b92c11.png">

### Acceptance Criteria
- [ ] External assets like jquery that we use in multiple themes should only be loaded once
- [ ] avoid manual configuration; e.g., if we add a new dependency, it would be nice for it to automatically go in the appropriate chunk.

**Comment by ChristopherChudzicki** [comment]
tldr: We should use `optimization.splitChunks` to automatically share code for us, not a manually made `main.js`. And we should configure `splitChunks` to ONLY consider "course-v2" and "www" when doing the splitting. That way experimenting with other themes won't affect our production bundles.

### Webpack terminology:
My understanding:
- **entrypoint:** File or files that tell Webpack where to start when following dependency graph to create bundles.
- **bundle:** A collection of files, the final output of webpack. One bundle per entrypoint. 
- **output chunk:** I will use this to mean an individual output file in the bundle *chunk is a webpack-specific term; it also refers to the internal representation of modules while webpack runs. I'll try to say "output chunk".*

The approach we've taken in OCW to sharing common assets between the course theme and the www theme has been to manually specify the shared code in `base-theme/assets/index.ts`, then use multiple entry points:
```js
{
  entry: {
    main:      "/base-theme/assets/index.ts",     //we manually put shared code here
    course_v2: "/course-v2/assets/course-v2.ts",
    www:       ".www/assets/www.tsx"
  }
// plus some other bundles for "fields
}
```
then we load two bundles on each page, "main+www" for www and "main+course_v2" for course sites. 

**Disadvantages:** This has the obvious disadvantage that if we accidentally import (or [provide](https://github.com/mitodl/ocw-hugo-themes/blob/main/base-theme/assets/webpack/webpack.common.ts#L139), in the case of jQuery and Popper) then the dependency in two different entrypoints then it will be included twice.

##  Auto(ish) asset sharing with`optimization.splitChunks` (aka SplitChunksPlugin)
Webpack's SplitChunksPlugin can handle most of this 'efficient asset sharing' business for us, putting common modules in output chunks that are shared between bundles.

How exactly webpack distributes modules among shared output chunks is pretty configurable. For example, you can configure webpack to do something like
```
### SplitChunksPlugin Example
Scenario:     three entrypoints e1, e2, e3
config:       create a new shared chunk whenever at least two entrypoints
                share a module
result:
    bundle e1 requires:
        e1.js 
        shared_e1_e2_e3.js
        shared_e1_e2.js
        shared_e1_e3.js
    etc
```
**Downside:** If you add a new entrypoint e4, you might get several new chunks (`e1_e4.js`, etc), and now you need to go update your `<script />` tags.

In some scenarios, `HtmlWebpackPlugin` can generate the appropriate script tags for you. I'm fairly certain that won't play well with Hugo templates, though. With some effort, you can make [`WebpackManifestPlugin`](https://www.npmjs.com/package/webpack-manifest-plugin) generate a list of files for each entrypoint.

### Some Options and a suggestion
So, how should we configure `optimization.splitChunks`?

We no longer need the "main" entrypoint: `optimization.splitChunks` is handling that for us. So we have:
```
{
  entry: {
    course_v2: "/course-v2/assets/course-v2.ts",
    course: "/course-v2/assets/course.ts",        // no longer used, but useful for discussion
    fields: "/fields/assets/course.ts",        // not active
    www:       ".www/assets/www.tsx"
  }
```
Note that I've included "course" and "fields" here. We don't actively use these, but they are useful for discussion.

Options:
1. **Single shared chunk `common_coursev2_www.js`** only for sharing between course-v2 and www
    - **Pros:** Simple script tags, all shareable code is shared, and it's stable: if we add other entrypoints (e.g., experimenting with course-v3) we don't affect production.
    - **Cons:** If we add a new theme with a new entrypoint, it won't automatically benefit from splitChunks
2. **Single shared chunk for all entrypoints:** Whenever *any 2+* entrypoints use the same module, put that module in a shared chunk `common.js`
    - **Pro:** Simple; it's [what we do in mitxonline](https://github.com/mitodl/mitxonline/blob/main/frontend/public/webpack.config.js#L95-L105)
    - **Cons:** It can be harmful to sizes. E.g., because "course" and "course-v2" both use `videojs`, `videojs` would go in the shared bundle and be downloaded by on `www` homepage. *Of course, we're about to get rid of "course", but we could have a similar scenario in future* 
3. **Let webpack generate chunks more dynamically:** similar to "SplitChunksPlugin Example" above
    - **Pros:** lot's of code sharing, even between new/future entrypoints
    - **Cons:** Generated filenames make `<script />` tags hard. We'd probably want to do that programmatically. *Probably not that hard—ManifestPlugin can get the dependencies [with some custom config](https://stackoverflow.com/a/69331449/2747370) but I don't think there's any builtin way to do it.*

IMO, Option 1 is the best bet.

**Comment by ChristopherChudzicki** [comment]
Completed by https://github.com/mitodl/ocw-hugo-themes/pull/1059
